dvc/repo/ls.py

```diff
@@ -0,0 +1,107 @@
+import os
+
+from dvc.exceptions import DvcException, NotDvcRepoError, OutputNotFoundError
+from dvc.scm.base import CloneError
+
+
+@staticmethod
+def ls(url, target=None, rev=None, recursive=None, outputs_only=False):
+    if os.path.exists(url):
+        if rev:
+            raise DvcException("Local repo doesn't support revisions")
+        return _ls_local_dir(url, target, recursive, outputs_only)
+
+    try:
+        return _ls_remote_git_repo(url, target, rev, recursive, outputs_only)
+    except CloneError:
+        raise DvcException("There is no such url {}".format(url))
+
+
+def _ls_files(path, target=None, recursive=None):
+    from dvc.scm.tree import WorkingTree
+    from dvc.ignore import CleanTree
+
+    def prettify(prefix, filename):
+        if not prefix == os.curdir:
+            filename = os.path.join(prefix, filename)
+        return filename
+
+    files = []
+    target_path = os.path.join(path, target) if target else path
+    tree = CleanTree(WorkingTree(target_path))
+
+    try:
+        target_path = os.path.join(path, target) if target else path
+        if not os.path.exists(target_path):
+            raise FileNotFoundError("{} doesn't exist".format(target_path))
+
+        if target:
+            files.append(prettify(os.curdir, target))
+            if not recursive:
+                return files
+
+        for dirpath, dirnames, filenames in tree.walk(target_path):
+            prefix = os.path.relpath(dirpath, path)
+            files.extend(map(lambda f: prettify(prefix, f), dirnames))
+            files.extend(map(lambda f: prettify(prefix, f), filenames))
+            if not recursive:
+                break
+    except (NotADirectoryError, FileNotFoundError):
+        pass
+
+    return files
+
+
+def _ls_outs_local_repo(repo, target=None, recursive=None):
+    def prettify(out):
+        relpath = out.path_info.relpath(repo.root_dir)
+        parts = list(relpath.parts)
+        segments = [parts.pop(0)]
+        if not recursive:
+            return segments
+
+        while parts:
+            segments.append(os.path.join(segments[-1], parts.pop(0)))
+        return segments
+
+    files = set()
+    try:
+        target_path = (
+            os.path.join(repo.root_dir, target) if target else repo.root_dir
+        )
+        outs = repo.find_outs_by_path(target_path, recursive=True)
+    except OutputNotFoundError:
+        outs = []
+
+    for out in outs:
+        if not recursive and target and len(outs):
+            files.add(target)
+            break
+        files.update(prettify(out))
+
+    return list(files)
+
+
+def _ls_local_dir(path, target=None, recursive=None, outputs_only=False):
+    from dvc.repo import Repo
+
+    result = set()
+    if not outputs_only:
+        result.update(_ls_files(path, target, recursive))
+    try:
+        repo = Repo(path)
+        result.update(_ls_outs_local_repo(repo, target, recursive))
+        repo.close()
```

Comment:
You are still not handling the closure correctly. This should be in the `finally` block, so if any error happens it will be closed.